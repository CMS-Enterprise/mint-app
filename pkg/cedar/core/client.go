package cedarcore

import (
	"context"
	"net/http"
	"time"

	ld "gopkg.in/launchdarkly/go-server-sdk.v5"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	cache "github.com/patrickmn/go-cache"
	"go.uber.org/zap"

	"github.com/cmsgov/easi-app/pkg/appcontext"
	apiclient "github.com/cmsgov/easi-app/pkg/cedar/core/gen/client"
	"github.com/cmsgov/easi-app/pkg/flags"
)

const (
	cedarCoreEnabledKey     = "cedarCoreEnabled"
	cedarCoreEnabledDefault = false
)

// NewClient builds the type that holds a connection to the CEDAR Core API
func NewClient(ctx context.Context, cedarHost string, cedarAPIKey string, ldClient *ld.LDClient) *Client {
	fnEmit := func(ctx context.Context) bool {
		lduser := flags.Principal(ctx)
		result, err := ldClient.BoolVariation(cedarCoreEnabledKey, lduser, cedarCoreEnabledDefault)
		if err != nil {
			appcontext.ZLogger(ctx).Info(
				"problem evaluating feature flag",
				zap.Error(err),
				zap.String("flagName", cedarCoreEnabledKey),
				zap.Bool("flagDefault", cedarCoreEnabledDefault),
				zap.Bool("flagResult", result),
			)
		}
		return result
	}

	c := cache.New(cache.NoExpiration, cache.NoExpiration) // Don't expire data _or_ clean it up

	hc := http.DefaultClient

	client := &Client{
		cedarCoreEnabled: fnEmit,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				apiclient.DefaultBasePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc:    hc,
		cache: c,
	}

	// Start cache refresh for systems
	client.startCacheRefresh(ctx, time.Minute*5, client.populateSystemSummaryCache)

	return client
}

// Client represents a connection to the CEDAR Core API
type Client struct {
	cedarCoreEnabled func(context.Context) bool
	auth             runtime.ClientAuthInfoWriter
	sdk              *apiclient.CEDARCoreAPI
	hc               *http.Client
	cache            *cache.Cache
}

// startCacheRefresh starts a goroutine that will run `populateCache` based on cacheRefreshTime
// This function returns no errors, and only logs when something goes wrong
func (c *Client) startCacheRefresh(ctx context.Context, cacheRefreshTime time.Duration, populateCache func(context.Context) error) {
	ticker := time.NewTicker(cacheRefreshTime)
	go func(ctx context.Context) {
		for {
			err := populateCache(ctx)
			if err != nil {
				appcontext.ZLogger(ctx).Error("Failed to refresh cache", zap.Error(err))
			}
			// Wait for the ticker. This will block the current goroutine until the ticker sends a message over the channel
			<-ticker.C
		}
	}(ctx)
}
