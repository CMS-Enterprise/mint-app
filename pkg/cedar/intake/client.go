// Package intake serves as an Anti-Corruption Layer (in Domain Driven Design
// parlance), keeping the implementation details of working with the
// downstream CEDAR Intake API well separated, so the rest of the internal
// workings of the EASi code are not using anything defined by the
// external API
package intake

import (
	"context"
	"fmt"
	"net/http"
	"strconv"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	"go.uber.org/zap"
	ld "gopkg.in/launchdarkly/go-server-sdk.v5"

	"github.com/cmsgov/easi-app/pkg/appcontext"
	apiclient "github.com/cmsgov/easi-app/pkg/cedar/intake/gen/client"
	"github.com/cmsgov/easi-app/pkg/cedar/intake/gen/client/health_check"
	"github.com/cmsgov/easi-app/pkg/cedar/intake/gen/client/intake"
	"github.com/cmsgov/easi-app/pkg/cedar/intake/translation"
	"github.com/cmsgov/easi-app/pkg/flags"
	"github.com/cmsgov/easi-app/pkg/models"
)

const (
	emitFlagKey = "emit-to-cedar"
	emitDefault = false
)

// NewClient builds the type that holds a connection to the CEDAR Intake API
func NewClient(cedarHost string, cedarAPIKey string, ldClient *ld.LDClient) *Client {
	fnEmit := func(ctx context.Context) bool {
		// this is the conditional way of stopping us from submitting to CEDAR; see EASI-1025
		lduser := flags.Principal(ctx)
		result, err := ldClient.BoolVariation(emitFlagKey, lduser, emitDefault)
		if err != nil {
			appcontext.ZLogger(ctx).Info(
				"problem evaluating feature flag",
				zap.Error(err),
				zap.String("flagName", emitFlagKey),
				zap.Bool("flagDefault", emitDefault),
				zap.Bool("flagResult", result),
			)
		}
		return result
	}

	hc := http.DefaultClient

	return &Client{
		emitToCedar: fnEmit,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				apiclient.DefaultBasePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc: hc,
	}
}

// Client represents a connection to the CEDAR Intake API
type Client struct {
	emitToCedar func(context.Context) bool
	auth        runtime.ClientAuthInfoWriter
	sdk         *apiclient.CEDARIntake
	hc          *http.Client
}

// CheckConnection hits the CEDAR Intake API `/healthcheck` endpoint to verify
// that our connection is functional
func (c *Client) CheckConnection(ctx context.Context) error {
	if !c.emitToCedar(ctx) {
		return nil
	}

	params := health_check.NewHealthCheckParamsWithContext(ctx)
	params.HTTPClient = c.hc

	resp, err := c.sdk.HealthCheck.HealthCheck(params, c.auth)
	if err != nil {
		return err
	}
	if resp.Payload == nil {
		return fmt.Errorf("no response body received")
	}
	return nil
}

// PublishSystemIntake sends a system intake to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishSystemIntake(ctx context.Context, si models.SystemIntake) error {
	intakeObject := translation.TranslatableSystemIntake(si)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishBusinessCase sends a business case to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishBusinessCase(ctx context.Context, bc models.BusinessCase) error {
	intakeObject := translation.TranslatableBusinessCase(bc)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishGRTFeedback sends an item of GRT feedback to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishGRTFeedback(ctx context.Context, feedback models.GRTFeedback) error {
	intakeObject := translation.TranslatableFeedback(feedback)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishAction sends an action to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishAction(ctx context.Context, action models.Action) error {
	intakeObject := translation.TranslatableAction(action)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishNote sends a note to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishNote(ctx context.Context, note models.Note) error {
	intakeObject := translation.TranslatableNote(note)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// private method for publishing anything that satisfies the translation.IntakeObject interface to CEDAR through the Intake API
func (c *Client) publishIntakeObject(ctx context.Context, model translation.IntakeObject) error {
	// constant values for now; may become non-constant if/when we revisit handling CEDAR validation errors

	// from the Swagger (cedar_intake.json, definitions/IntakeInput/properties/version):
	// "The version associated with the object in the body. This value can be incremented in the event a transaction needs to be resubmitted."
	const objectVersion = 1

	// from the Swagger (cedar_intake.json, paths/"/intake"/post/parameters/validatePayload):
	// "Determines if schema validation of the payload is performed syncronously before persisting the record or asyncronously after the record has been persisted"
	const isValidatedSynchronously = false // false for V1 schemas; this may change when we implement V2 schemas for EASI-1614

	id := model.ObjectID()
	objectType := model.ObjectType()

	if !c.emitToCedar(ctx) {
		appcontext.ZLogger(ctx).Info("snapshot publishing disabled", zap.String("object ID", id), zap.String("object type", objectType))
		return nil
	}

	input, err := model.CreateIntakeModel()

	if err != nil {
		appcontext.ZLogger(ctx).Error(
			"problem building cedar payload",
			zap.String("object ID", id),
			zap.String("object type", objectType),
			zap.Error(err),
		)
		return nil
	}

	params := intake.NewIntakeAddParamsWithContext(ctx)
	params.HTTPClient = c.hc
	params.Body = input

	params.ValidatePayload = strconv.FormatBool(isValidatedSynchronously)

	objectVersionStr := strconv.FormatInt(objectVersion, 10)
	params.Body.Version = &objectVersionStr

	resp, err := c.sdk.Intake.IntakeAdd(params, c.auth)
	if err != nil {
		return err
	}

	// TODO: we may need to read through the results for any "duplicate entry" error codes
	// that CEDAR _may_ send us, especially if we are re-transmitting entities... We
	// don't really care if there are dupes on their side, because we do NOT want to hold
	// any state on our side about who/what/when may have transmitted a previous copy
	// of the info to CEDAR.
	// appcontext.ZLogger(ctx).Info(fmt.Sprintf("results: %v", resp.Payload))
	_ = resp

	return nil
}
