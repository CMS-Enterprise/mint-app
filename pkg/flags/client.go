package flags

import (
	"context"
	"crypto/sha256"
	"fmt"
	"time"

	"github.com/cmsgov/mint-app/pkg/appcontext"

	"gopkg.in/launchdarkly/go-sdk-common.v2/lduser"
	ld "gopkg.in/launchdarkly/go-server-sdk.v5"
	"gopkg.in/launchdarkly/go-server-sdk.v5/ldfiledata"
	"gopkg.in/launchdarkly/go-server-sdk.v5/ldfilewatch"

	"github.com/cmsgov/mint-app/pkg/appconfig"
)

// DowngradeAssessmentTeamKey is the Flag Key in LaunchDarkly for downgrading assessment team users
const DowngradeAssessmentTeamKey = "downgradeAssessmentTeam"

// Config has all the parts for creating a new LD Client
type Config struct {
	Source         appconfig.FlagSourceOption
	Key            string
	Timeout        time.Duration
	FlagValuesFile string
}

// NewLaunchDarklyClient returns a client backed by Launch Darkly
func NewLaunchDarklyClient(config Config) (*ld.LDClient, error) {
	if config.Source == appconfig.FlagSourceLaunchDarkly {
		return ld.MakeClient(config.Key, config.Timeout)
	}

	defaultTimeout := 5 * time.Second

	if config.Source == appconfig.FlagSourceFile {

		ldConfig := ld.Config{
			DataSource: ldfiledata.DataSource().FilePaths(config.FlagValuesFile).Reloader(ldfilewatch.WatchFiles),
		}
		return ld.MakeCustomClient("fake_key", ldConfig, defaultTimeout)
	}

	// we default to an OFFLINE client for non-deployed environments
	return ld.MakeCustomClient("fake_offline_key", ld.Config{Offline: true}, 5*time.Second)
}

// Principal builds the LaunchDarkly user object for the
// currently authenticated principal.
func Principal(ctx context.Context) lduser.User {
	p := appcontext.Principal(ctx)
	key := UserKeyForID(p.ID())

	// this is a bit of a loose inference, assuming a user w/o Job Codes
	// is an Anonymous user. Over time, may want to consider adding
	// a `func Anonymous() bool` accessor to the authorizaion.Principal interface
	// definition instead of doing this inference
	authed := (p.AllowUSER() || p.AllowASSESSMENT())

	return lduser.
		NewUserBuilder(key).
		Anonymous(!authed).
		Build()
}

// UserKeyForID generates a user key from an ID
// we should not be using bare EUA IDs as identifiers to
// LaunchDarkly (per Jimil/ISSO), so we use a cryptographically
// secure one-way hash of the EUA ID as "key" for the LD User object.
func UserKeyForID(id string) string {
	h := sha256.New()
	_, _ = h.Write([]byte(id))
	return fmt.Sprintf("%x", h.Sum(nil))
}
