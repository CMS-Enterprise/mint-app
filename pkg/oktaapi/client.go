package oktaapi

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/okta/okta-sdk-golang/v2/okta"
	"github.com/okta/okta-sdk-golang/v2/okta/query"

	"github.com/cmsgov/mint-app/pkg/models"
)

// Client is a wrapper around github.com/okta/okta-sdk-golang/v2/okta Client type.
// The purpose of this package is to act as a drop-in replacement for the CEDAR LDAP API, so this package should implement the same
// methods that the Client interface defines in pkg/cedar/cedarldap/translated_client.go
type Client struct {
	oktaClient *okta.Client
}

// NewClient creates a Client
func NewClient(url string, token string) (*Client, error) {
	// TODO Do we need the "Context" response from okta.NewClient??
	_, oktaClient, oktaClientErr := okta.NewClient(context.TODO(), okta.WithOrgUrl(url), okta.WithToken(token))
	if oktaClientErr != nil {
		return nil, oktaClientErr
	}
	return &Client{
		oktaClient: oktaClient,
	}, nil
}

type oktaUserProfile struct {
	FirstName   string `json:"firstName"`
	LastName    string `json:"lastName"`
	DisplayName string `json:"displayName"`
	Email       string `json:"email"`
	Login       string `json:"login"`
}

func parseOktaProfileResponse(profile *okta.UserProfile) (*models.UserInfo, error) {
	// Create an okaUserProfile to return
	parsedProfile := &oktaUserProfile{}

	// Marshal the profile into a string so we can later unmarshal it into a struct
	responseString, err := json.Marshal(profile)
	if err != nil {
		return nil, err
	}

	// Unmarshal the string into the oktaUserProfile type
	json.Unmarshal(responseString, parsedProfile)

	returnInfo := &models.UserInfo{
		CommonName: parsedProfile.DisplayName,
		Email:      models.EmailAddress(parsedProfile.Email),
		EuaUserID:  parsedProfile.Login,
		FirstName:  parsedProfile.FirstName,
		LastName:   parsedProfile.LastName,
	}

	return returnInfo, nil
}

func (o *Client) FetchUserInfo(ctx context.Context, username string) (*models.UserInfo, error) {
	user, _, err := o.oktaClient.User.GetUser(ctx, username)
	if err != nil {
		return nil, err
	}

	profile, err := parseOktaProfileResponse(user.Profile)
	if err != nil {
		return nil, err
	}

	return profile, nil
}

func (o *Client) SearchCommonNameContains(ctx context.Context, searchTerm string) ([]*models.UserInfo, error) {
	filterString := fmt.Sprintf(`profile.firstName eq "%v"`, searchTerm)
	filter := query.NewQueryParams(query.WithFilter(filterString))

	filteredUsers, _, err := o.oktaClient.User.ListUsers(ctx, filter)
	if err != nil {
		fmt.Printf("Error Getting Users: %v\n", err)
	}

	users := make([]*models.UserInfo, len(filteredUsers))
	for idx, user := range filteredUsers {
		profile, err := parseOktaProfileResponse(user.Profile)
		if err != nil {
			return nil, err
		}
		users[idx] = profile
	}
	return users, nil
}
