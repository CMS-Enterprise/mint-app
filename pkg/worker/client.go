package worker

import (
	"fmt"

	faktory "github.com/contribsys/faktory/client"
)

// NewClient returns a new faktory Client.
func NewClient() (*faktory.Client, error) {
	cl, err := faktory.Open()
	if err != nil {
		return nil, fmt.Errorf("error opening faktory client: %w", err)
	}
	return cl, nil
}
