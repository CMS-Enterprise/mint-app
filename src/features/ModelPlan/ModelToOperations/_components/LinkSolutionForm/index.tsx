import React, {
  Dispatch,
  SetStateAction,
  useCallback,
  useEffect,
  useMemo,
  useState
} from 'react';
import { Trans, useTranslation } from 'react-i18next';
import { useParams } from 'react-router-dom';
import { Grid, GridContainer, Label, Link } from '@trussworks/react-uswds';
import { helpSolutions } from 'features/HelpAndKnowledge/SolutionsHelp/solutionsMap';
import {
  GetMtoAllSolutionsQuery,
  GetMtoMilestoneQuery,
  MtoCommonSolutionKey
} from 'gql/generated/graphql';

import HelpText from 'components/HelpText';
import UswdsReactLink from 'components/LinkWrapper';
import MultiSelect from 'components/MultiSelect';

import { SolutionCard } from '../SolutionCard';

import '../../index.scss';

type LinkSolutionFormProps = {
  milestone: GetMtoMilestoneQuery['mtoMilestone'];
  commonSolutionKeys: MtoCommonSolutionKey[];
  setCommonSolutionKeys: Dispatch<SetStateAction<MtoCommonSolutionKey[]>>;
  solutionIDs: string[];
  setSolutionIDs: Dispatch<SetStateAction<string[]>>;
  allSolutions: GetMtoAllSolutionsQuery['modelPlan']['mtoMatrix'];
};

const LinkSolutionForm = ({
  milestone,
  commonSolutionKeys,
  setCommonSolutionKeys,
  solutionIDs,
  setSolutionIDs,
  allSolutions
}: LinkSolutionFormProps) => {
  const { t } = useTranslation('modelToOperationsMisc');
  const { t: milestoneT } = useTranslation('mtoMilestone');

  // console.log('commonSolutionKeys', commonSolutionKeys);
  // console.log('solutionIDs', solutionIDs);

  const { modelID } = useParams<{ modelID: string }>();

  // Map the common solutions to the FE help solutions
  const mappedSolutions = useMemo(
    () =>
      milestone.commonMilestone?.commonSolutions.map(solution => {
        return helpSolutions.find(s => s.enum === solution.key);
      }) || [],
    [milestone.commonMilestone]
  );

  const mappedSolutionKeys = mappedSolutions.map(solution => solution?.enum);

  const commonSolutions =
    allSolutions?.commonSolutions.filter(
      solution => !mappedSolutionKeys.includes(solution.key)
    ) || [];

  const createdSolutions = useMemo(
    () => allSolutions?.solutions?.filter(solution => !solution.key) || [],
    [allSolutions?.solutions]
  );

  const groupedOptions = [
    {
      label: 'Custom solutions added to this MTO',
      options: createdSolutions.map(solution => {
        return {
          label: solution.name || '',
          value: solution.id
        };
      })
    },
    {
      label: 'Other available solutions',
      options: commonSolutions.map(solution => {
        return {
          label: solution.name || '',
          value: solution.key
        };
      })
    }
  ];

  const initialValues = [...commonSolutionKeys, ...solutionIDs];

  const isCustomSolution = useCallback(
    (id: string) => {
      return createdSolutions.find(solution => solution.id === id);
    },
    [createdSolutions]
  );

  const isSuggestedSolution = useCallback(
    (key: string) => {
      return mappedSolutionKeys.find(k => k === key);
    },
    [mappedSolutionKeys]
  );

  const [selectedSolutions, setSelectedSolutions] =
    useState<string[]>(initialValues);

  const setChecked = (key: MtoCommonSolutionKey) => {
    if (commonSolutionKeys.includes(key)) {
      setCommonSolutionKeys(commonSolutionKeys.filter(k => k !== key));
    } else {
      setCommonSolutionKeys([...commonSolutionKeys, key]);
    }
  };

  useEffect(() => {
    const custom: string[] = [];
    const common: MtoCommonSolutionKey[] = [];

    selectedSolutions.forEach(solution => {
      if (isCustomSolution(solution)) {
        custom.push(solution);
      } else if (!isSuggestedSolution(solution)) {
        common.push(solution as MtoCommonSolutionKey);
      }
    });

    const suggestedSolutions = mappedSolutions
      .filter(s => commonSolutionKeys.includes(s?.enum as MtoCommonSolutionKey))
      .map(s => s?.enum) as MtoCommonSolutionKey[];

    setSolutionIDs(custom);
    setCommonSolutionKeys([
      ...(common as unknown as MtoCommonSolutionKey[]),
      ...suggestedSolutions
    ]);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    selectedSolutions,
    isCustomSolution,
    setCommonSolutionKeys,
    setSolutionIDs,
    mappedSolutions
  ]);

  const selectedSolutionCount: number =
    commonSolutionKeys.length + solutionIDs.length;

  return (
    <GridContainer className="padding-8">
      <Grid row>
        <Grid col={12}>
          <h3 className="margin-bottom-4">
            {t('modal.editMilestone.selectedSolutionCount', {
              count: selectedSolutionCount
            })}
          </h3>

          <div className="border-bottom-1px border-base-lighter border-top-1px padding-top-4 padding-bottom-2 margin-bottom-4">
            <h4 className="margin-0">
              {t('modal.editMilestone.suggestedSolutions')}
            </h4>

            <p className="margin-top-0 margin-bottom-3 mint-body-normal text-base">
              {t('modal.editMilestone.selectedSolutionsDescription')}
            </p>

            {mappedSolutions.map(solution =>
              solution ? (
                <SolutionCard
                  key={solution.key}
                  solution={solution}
                  setChecked={setChecked}
                  checked={commonSolutionKeys.includes(
                    solution.enum as MtoCommonSolutionKey
                  )}
                />
              ) : null
            )}
          </div>

          <Label htmlFor="available-solutions">
            {milestoneT('solutions.label')}
          </Label>

          <HelpText className="margin-top-1">
            <Trans
              i18nKey="modelToOperationsMisc:modal.editMilestone.availableSolutionsDescription"
              components={{
                solution: (
                  <UswdsReactLink
                    to={`/models/${modelID}/collaboration-area/model-to-operations/solution-library`}
                  >
                    {' '}
                  </UswdsReactLink>
                ),
                help: (
                  <Link
                    href="/help-and-knowledge/operational-solutions"
                    target="_blank"
                    variant="external"
                  >
                    {' '}
                  </Link>
                )
              }}
            />
          </HelpText>

          <MultiSelect
            id="available-solutions"
            inputId="available-solutions"
            ariaLabel="Available solutions"
            options={[]}
            groupedOptions={groupedOptions}
            selectedLabel="Selected solutions"
            initialValues={selectedSolutions}
            name="availableSolutions"
            onChange={values => {
              setSelectedSolutions(values);
            }}
          />
        </Grid>
      </Grid>
    </GridContainer>
  );
};

export default LinkSolutionForm;
