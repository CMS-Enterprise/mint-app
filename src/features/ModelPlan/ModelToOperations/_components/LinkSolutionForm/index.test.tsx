import React from 'react';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import { render, screen } from '@testing-library/react';
import {
  GetMtoMilestoneQuery,
  MtoCommonMilestoneKey,
  MtoCommonSolutionKey,
  MtoMilestoneStatus,
  MtoRiskIndicator,
  MtoSolutionStatus
} from 'gql/generated/graphql';
import {
  categoryMock,
  milestoneMock,
  modelID,
  possibleSolutionsMock
} from 'tests/mock/mto';

import MessageProvider from 'contexts/MessageContext';

import LinkSolutionForm from '.';

type MilestoneType = GetMtoMilestoneQuery['mtoMilestone'];

const milestone: MilestoneType = {
  __typename: 'MTOMilestone',
  id: '123',
  name: 'Milestone 1',
  key: MtoCommonMilestoneKey.ACQUIRE_AN_EVAL_CONT,
  responsibleComponent: [],
  facilitatedBy: [],
  needBy: '2021-08-01',
  status: MtoMilestoneStatus.COMPLETED,
  riskIndicator: MtoRiskIndicator.AT_RISK,
  addedFromMilestoneLibrary: true,
  isDraft: false,
  categories: {
    __typename: 'MTOCategories',
    category: {
      __typename: 'MTOCategory',
      id: '1',
      name: 'Category 1'
    },
    subCategory: {
      __typename: 'MTOSubcategory',
      id: '2',
      name: 'Subcategory 1'
    }
  },
  commonMilestone: {
    __typename: 'MTOCommonMilestone',
    description: 'Description 1',
    key: MtoCommonMilestoneKey.ACQUIRE_AN_EVAL_CONT,
    commonSolutions: [
      {
        __typename: 'MTOCommonSolution',
        key: MtoCommonSolutionKey.BCDA
      }
    ]
  },
  solutions: [
    {
      __typename: 'MTOSolution',
      id: '1',
      name: 'Solution 1',
      key: MtoCommonSolutionKey.BCDA,
      status: MtoSolutionStatus.COMPLETED,
      riskIndicator: MtoRiskIndicator.AT_RISK,
      commonSolution: {
        __typename: 'MTOCommonSolution',
        name: 'common solution 1',
        key: MtoCommonSolutionKey.BCDA,
        isAdded: true
      }
    }
  ],
  notes: []
};

describe('LinkSolutionForm', () => {
  it('renders and matches snapshot', async () => {
    const router = createMemoryRouter(
      [
        {
          path: '/models/:modelID/collaboration-area/model-to-operations/matrix',
          element: (
            <MessageProvider>
              <LinkSolutionForm
                milestone={milestone}
                commonSolutionKeys={[MtoCommonSolutionKey.BCDA]}
                setCommonSolutionKeys={() => null}
                solutionIDs={[]}
                setSolutionIDs={() => null}
                allSolutions={{
                  __typename: 'ModelsToOperationMatrix',
                  commonSolutions: [],
                  info: {
                    __typename: 'MTOInfo',
                    id: '123'
                  },
                  solutions: []
                }}
                setCloseDestination={() => null}
              />
            </MessageProvider>
          )
        }
      ],
      {
        initialEntries: [
          `/models/${modelID}/collaboration-area/model-to-operations/matrix?view=milestones&edit-milestone=123`
        ]
      }
    );

    const { asFragment } = render(
      <MockedProvider
        mocks={[
          ...milestoneMock('123'),
          ...categoryMock,
          ...possibleSolutionsMock
        ]}
        addTypename={false}
      >
        <RouterProvider router={router} />
      </MockedProvider>
    );

    expect(screen.getByRole('checkbox')).toBeChecked();

    expect(asFragment()).toMatchSnapshot();
  });
});
