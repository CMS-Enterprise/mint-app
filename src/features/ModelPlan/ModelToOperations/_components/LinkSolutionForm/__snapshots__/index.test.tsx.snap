// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LinkSolutionForm > renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid-container padding-8"
    data-testid="gridContainer"
  >
    <div
      class="grid-row"
      data-testid="grid"
    >
      <div
        class="grid-col-10"
        data-testid="grid"
      >
        <h3
          class="margin-bottom-2"
        >
          1 selected solution
        </h3>
        <div
          class="text-base-dark mint-body-normal margin-bottom-4"
        >
          You may visit the 
          <button
            class="usa-button usa-button--unstyled usa-button--unstyled margin-0"
            data-testid="button"
            type="button"
          >
            solution library
          </button>
           or 
          <a
            class="usa-link usa-link--external"
            href="/help-and-knowledge/operational-solutions"
            target="_blank"
          >
            help center
          </a>
           to learn more about the solutions below.
        </div>
        <div
          class="border-bottom-1px border-base-lighter border-top-1px padding-top-4 padding-bottom-2 margin-bottom-4"
        >
          <h4
            class="margin-0"
          >
            Suggested solutions
          </h4>
          <p
            class="margin-top-0 margin-bottom-3 mint-body-normal text-base"
          >
            These solutions are commonly used for this milestone.
          </p>
          <div
            class="desktop:grid-col-9 display-flex"
            data-testid="grid"
          >
            <li
              class="usa-card width-full"
              data-testid="Card"
            >
              <div
                class="usa-card__container padding-3 flex-justify radius-md margin-0 model-to-operations__selected-shadow"
              >
                <div
                  class="mint-checkbox usa-checkbox"
                >
                  <input
                    checked=""
                    class="usa-checkbox__input"
                    id="bcda"
                    name="bcda"
                    type="checkbox"
                    value="bcda"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="bcda"
                  >
                    Select this solution
                  </label>
                  <p
                    class="margin-y-1 margin-left-4 line-height-body-3 font-body-2xs text-pre-line"
                  />
                </div>
                <h3
                  class="margin-0"
                >
                  Beneficiary Claims Data API
                </h3>
                <p
                  class="margin-top-0 margin-bottom-0"
                >
                  BCDA
                </p>
              </div>
            </li>
          </div>
        </div>
        <label
          class="usa-label"
          data-testid="label"
          for="available-solutions"
        >
          Available solutions
        </label>
        <div
          class="mint-help-text margin-top-1 text-base-dark mint-body-normal"
        >
          Select from other operational solutions and IT systems included in MINT, including any custom solutions you may have added to your MTO. Select all that apply.
        </div>
        <div>
          <div
            class="easi-multiselect usa-combo-box margin-top-1 grouped-options css-b62m3t-container"
            id="available-solutions"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-2-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              class=" css-1vl89cy-control"
            >
              <div
                class=" css-1fdsijx-ValueContainer"
              >
                <div
                  class=" css-1jqq78o-placeholder"
                  id="react-select-2-placeholder"
                >
                  0 selected
                </div>
                <div
                  class=" css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-activedescendant=""
                    aria-autocomplete="list"
                    aria-describedby="react-select-2-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    aria-labelledby="Available solutions"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class=""
                    id="available-solutions"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class=" css-16b81-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class=" css-gjng1h-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div>
              <input
                name="availableSolutions"
                type="hidden"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
